x-pack/plugins/security/server/routes/api/v1/authenticate.js

```diff
@@ -31,9 +30,8 @@ export function initAuthenticateApi(server) {
       const { username, password } = request.payload;
 
       try {
-        const authenticationResult = await server.plugins.security.authenticate(
-          BasicCredentials.decorateRequest(request, username, password)
-        );
+        request.loginAttempt().setCredentials(username, password);
```

Comment:
suggestion: it might make this clearer to pass the credentials through as an optional second parameter to `server.plugins.security.authenticate`, as opposed to setting the LoginAttempt implicitly tied to the request using a `WeakMap`.

Reply:
That was essentially one of my original proposals, but when putting together the [original prototype](https://github.com/elastic/kibana/pull/25431), @azasypkin mentioned that he felt a request-based approach would be better than changing the public API of the authenticator to support a feature that is only actually relevant in the context of a subset of providers.

The `WeakMap` detail is an artifact of the way hapi handles these request-based capabilities that are added via plugins, specifically that they provide no mechanism to provide the services as singletons. With an alternative HTTP handler implementation, like we've proposed in the new platform, this would not be necessary at all.

Even though it is an implementation detail, I don't love the `WeakMap` approach. But I'm encouraged that it's limited only to the hapi integration itself and we know it has a finite timeline. I'm open to ideas here though if you'd like to tackle this in a different way.